src/pip/_internal/network/download.py

```diff
@@ -92,96 +121,301 @@ def parse_content_disposition(content_disposition: str, default_filename: str) -
+) -> Iterator[_ErrorReceiver]:
+    err_recv = _ErrorReceiver(error_flag)
+    try:
+        for w in workers:
+            w.start()
+            # We've sorted the list of worker threads so they correspond to the largest
+            # downloads first. Each thread immediately waits upon a semaphore to limit
+            # maximum parallel downloads, and we would like the semaphore's internal
+            # wait queue to retain the same order we established earlier (otherwise, we
+            # would end up nondeterministically downloading files out of our desired
+            # order). Yielding to the scheduler here is intended to give the thread we
+            # just started time to jump into the semaphore, either to execute further
+            # (until the semaphore is full) or to jump into the queue at the desired
+            # position. We seem to get the ordering reliably even without this explicit
+            # yield, and ideally we would like to somehow ensure this deterministically,
+            # but this is relatively idiomatic and lets us lean on much fewer
+            # synchronization constructs. We can revisit this if users find the ordering
+            # is unreliable. It's easy to see if we've messed up, as the rich progress
+            # table prominently shows each download size and which ones are executing.
+            time.sleep(0)
+        yield err_recv
+    except BaseException as e:
+        err_recv.receive_error(e)
+    finally:
+        thread_exception = err_recv.stored_error()
+        if thread_exception is not None:
+            logger.critical("Received exception, shutting down downloader threads...")
+
+        # Ensure each thread is complete by the time the queue has exited, either by
+        # writing the full request contents, or by checking the Event from an exception.
+        for w in workers:
+            # If the user (reasonably) wants to hit ^C again to try to make it close
+            # faster, we want to avoid spewing out a ton of error text, but at least
+            # let's let them know we hear them and we're trying to shut down!
+            while w.is_alive():
+                try:
+                    w.join()
+                except BaseException:
+                    logger.critical("Shutting down worker threads, please wait...")
+
+        if thread_exception is not None:
+            raise thread_exception
+
+
+def _copy_chunks(
+    output_queue: "Queue[Union[Tuple[Link, Path, Optional[str]], BaseException]]",
+    error_flag: Event,
+    semaphore: Semaphore,
+    session: PipSession,
+    location: Path,
+    batched_progress: BatchedProgress,
+    download_info: Tuple[Link, TaskID, str],
+) -> None:
+    link, task_id, filename = download_info
+
+    with semaphore:
+        # Check if another thread exited with an exception before we started.
+        if error_flag.is_set():
+            return
+        try:
+            resp = _http_get_download(session, link)
```

Comment:
Might be worth getting the raw socket from the request so the main thread can close it on a KeyboardInterrupt